src/main/kotlin/com/exactpro/th2/codec/grpc/GrpcCodecService.kt

```diff
@@ -0,0 +1,123 @@
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package com.exactpro.th2.codec.grpc
+
+import com.exactpro.th2.codec.CodecException
+import com.exactpro.th2.codec.StoreEventProcessor
+import com.exactpro.th2.codec.util.messageIds
+import com.exactpro.th2.common.utils.message.sessionAlias
+import com.exactpro.th2.common.event.Event
+import com.exactpro.th2.common.grpc.AnyMessage
+import com.exactpro.th2.common.grpc.EventID
+import com.exactpro.th2.common.grpc.MessageGroupBatch
+import com.exactpro.th2.common.schema.grpc.router.GrpcRouter
+import io.grpc.Status
+import io.grpc.stub.StreamObserver
+import mu.KotlinLogging
+
+class GrpcCodecService(
+    grpcRouter: GrpcRouter,
+    private val generalDecodeFunc: ((MessageGroupBatch) -> MessageGroupBatch),
+    private val generalEncodeFunc: ((MessageGroupBatch) -> MessageGroupBatch),
+    onEvent: (Event, EventID?) -> Unit,
+    private val isFirstCodecInPipeline: Boolean
+) : CodecGrpc.CodecImplBase() {
+
+    private val eventProcessor = StoreEventProcessor(onEvent)
+
+    private val nextCodec = try {
+        grpcRouter.getService(AsyncCodecService::class.java)
+    } catch (t: Throwable) {
+        null
+    }
+
+    private val MessageGroupBatch.sessionAlias: String
+        get() {
+            var sessionAlias: String? = null
+            for (group in groupsList) {
+                for (message in group.messagesList) {
+                    when (sessionAlias) {
+                        null -> sessionAlias = message.sessionAlias
+                        else -> require(sessionAlias == message.sessionAlias) { "Batch contains more than one session alias" }
+                    }
+                }
+            }
+            return requireNotNull(sessionAlias) { "Batch is empty" }
+        }
+
+    override fun decode(batch: MessageGroupBatch, responseObserver: StreamObserver<MessageGroupBatch>) {
+        try {
+            val parsed = generalDecodeFunc(batch)
+            if (parsed.anyMessage(AnyMessage::hasRawMessage)) {
+                if (nextCodec == null) throw CodecException("grpc codec pipeline output contains raw messages after decoding")
+                nextCodec.decode(parsed, mapOf("session_alias" to parsed.sessionAlias), responseObserver)
+            } else {
+                responseObserver.onNext(parsed)
+                responseObserver.onCompleted()
+            }
+        } catch (t: Throwable) {
```

Comment:
I believe that we should catch `Exception` here instead of `Throwable`. Otherwise, we might suppress OutOfMemory or something else critical